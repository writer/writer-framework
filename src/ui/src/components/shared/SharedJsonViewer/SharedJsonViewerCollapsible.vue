<template>
	<SharedCollapsible
		:open="open"
		:disabled="disabled"
		:icons="{ close: 'keyboard_arrow_right', open: 'keyboard_arrow_down' }"
		@toggle="$emit('toggle', $event)"
	>
		<template #title>
			<div class="SharedJsonViewerCollapsible__title">
				<span v-if="title">{{ title }}</span>
				<SharedJsonViewerChildrenCounter v-if="data" :data="data" />
			</div>
		</template>
		<template #content>
			<div class="SharedJsonViewerCollapsible__content">
				<slot />
			</div>
		</template>
	</SharedCollapsible>
</template>

<script setup lang="ts">
import type { PropType } from "vue";
import SharedCollapsible from "../SharedCollapsible.vue";
import type { JsonData } from "./SharedJsonViewer.vue";
import SharedJsonViewerChildrenCounter from "./SharedJsonViewerChildrenCounter.vue";

defineProps({
	open: { type: Boolean, required: false },
	disabled: { type: Boolean, required: false },
	title: { type: String, required: false, default: undefined },
	data: {
		type: [
			String,
			Number,
			Boolean,
			Object,
			Array,
			null,
		] as PropType<JsonData>,
		required: false,
		default: undefined,
	},
});

defineEmits({
	toggle: (open: boolean) => typeof open === "boolean",
});
</script>

<style scoped>
.SharedJsonViewerCollapsible__title {
	font-family: monospace;
	font-size: 12px;
	display: flex;
	gap: 8px;
}

.SharedJsonViewerCollapsible__content {
	margin-left: 7px;
	padding-left: var(--jsonViewerIndentationSpacing, 8px);
	padding-top: 4px;
	padding-bottom: 4px;
}
</style>
