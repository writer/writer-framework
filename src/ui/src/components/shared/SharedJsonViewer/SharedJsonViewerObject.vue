<template>
	<div class="SharedJsonViewerObject">
		<template
			v-for="([key, value], index) in Object.entries(data)"
			:key="index"
		>
			<!-- This is a single value, we display it as plain key/value text -->
			<div
				v-if="isJSONValue(value) || getJSONLength(value) === 0"
				class="SharedJsonViewerObject__value"
			>
				<span>{{ key }}</span
				>:
				<SharedJsonViewerValue :data="value" />
			</div>
			<!-- This is a an object -->
			<SharedJsonViewerCollapsible
				v-else
				:open="isOpen(key)"
				:title="key"
				:data="value"
				@toggle="toggleOpenedKey(key, $event)"
			>
				<SharedJsonViewer
					v-if="isOpen(key)"
					:data="value"
					:path="[...path, key]"
					:initial-depth="initialDepth"
					@toggle="$emit('toggle', $event)"
				/>
			</SharedJsonViewerCollapsible>
		</template>
	</div>
</template>

<script setup lang="ts">
import { PropType, computed, ref, toRef, watch } from "vue";
import {
	getJSONLength,
	isJSONValue,
	jsonViewerToggleEmitDefinition,
} from "./SharedJsonViewer.utils";
import type { JsonData, JsonPath } from "./SharedJsonViewer.vue";
import SharedJsonViewer from "./SharedJsonViewer.vue";
import SharedJsonViewerCollapsible from "./SharedJsonViewerCollapsible.vue";
import SharedJsonViewerValue from "./SharedJsonViewerValue.vue";

const props = defineProps({
	data: {
		type: [Object, Array] as PropType<JsonData>,
		required: true,
	},
	path: {
		type: Array as PropType<JsonPath>,
		default: () => [],
	},
	initialDepth: { type: Number, default: 0 },
});

const initialDepth = toRef(props, "initialDepth");

const emit = defineEmits({
	toggle: jsonViewerToggleEmitDefinition,
});

const openedKeys = ref<string[]>([]);

const currentLevel = computed(() => props.path.length + 1);

watch(
	initialDepth,
	() => {
		if (initialDepth.value === 0) {
			openedKeys.value = [];
			return;
		}

		const shouldAddKeys =
			initialDepth.value === -1 ||
			initialDepth.value > currentLevel.value;

		if (!shouldAddKeys) return;

		const keysToOpen = new Set([
			...openedKeys.value,
			...Object.keys(props.data),
		]);

		openedKeys.value = Array.from(keysToOpen);
	},
	{ immediate: true },
);

function toggleOpenedKey(key: string, open: boolean) {
	if (open) {
		openedKeys.value = [...openedKeys.value, key];
	} else {
		openedKeys.value = openedKeys.value.filter((k) => k !== key);
	}
	emit("toggle", { open, path: [...props.path, key] });
}

function isOpen(key: string) {
	return openedKeys.value.includes(String(key));
}
</script>

<style scoped>
.SharedJsonViewerObject {
	display: flex;
	flex-direction: column;
	gap: 4px;
}

.SharedJsonViewerObject__value {
	font-family: monospace;
	font-size: 12px;
	padding-left: 24px;

	display: flex;
	gap: 4px;
}
</style>
