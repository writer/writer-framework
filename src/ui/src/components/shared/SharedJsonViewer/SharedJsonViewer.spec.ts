import { describe, expect, it } from "vitest";
import SharedJsonViewer from "./SharedJsonViewer.vue";
import SharedJsonViewerValue from "./SharedJsonViewerValue.vue";
import SharedJsonViewerObject from "./SharedJsonViewerObject.vue";
import { mount } from "@vue/test-utils";

describe("SharedJsonViewer", () => {
	it("should render a string", () => {
		const wrapper = mount(SharedJsonViewer, {
			propsData: { data: "string" },
		});
		const jsonValue = wrapper.getComponent(SharedJsonViewerValue);
		expect(jsonValue.text()).toBe('"string"');
	});

	it("should render a boolean", () => {
		const wrapper = mount(SharedJsonViewer, { propsData: { data: true } });
		const jsonValue = wrapper.getComponent(SharedJsonViewerValue);
		expect(jsonValue.text()).toBe("true");
	});

	it("should render a null", () => {
		const wrapper = mount(SharedJsonViewer, { propsData: { data: null } });
		const jsonValue = wrapper.getComponent(SharedJsonViewerValue);
		expect(jsonValue.text()).toBe("null");
	});

	it("should render an object", () => {
		const data = { a: "a", b: "b" };
		const wrapper = mount(SharedJsonViewer, { propsData: { data } });
		const jsonObject = wrapper.getComponent(SharedJsonViewerObject);
		expect(jsonObject.props().data).toStrictEqual(data);

		expect(wrapper.findAllComponents(SharedJsonViewerValue)).toHaveLength(
			2,
		);
	});

	it("should render a nested object", async () => {
		const data = { array: [1, 2, 3], nested: { foo: "bar", a: { b: 2 } } };
		const wrapper = mount(SharedJsonViewer, { propsData: { data } });

		expect(wrapper.element).toMatchSnapshot();
	});

	it("should render an array", async () => {
		const data = [1, 2, 3];
		const wrapper = mount(SharedJsonViewer, { propsData: { data } });

		expect(wrapper.element).toMatchSnapshot();
	});
});
