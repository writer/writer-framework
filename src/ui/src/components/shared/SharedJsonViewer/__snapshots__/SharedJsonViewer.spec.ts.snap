// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SharedJsonViewer > should render a nested object 1`] = `
<div
  data-v-app=""
>
  
  
  <details
    class="SharedCollapsible"
    data-v-11af3d6a=""
    data-v-55c4d015=""
    disabled="false"
  >
    <summary
      class="SharedCollapsible__summary"
      data-v-55c4d015=""
    >
      <!--v-if-->
      
      <div
        class="SharedJsonViewerCollapsible__title"
        data-v-11af3d6a=""
      >
        <!--v-if-->
        <span
          data-v-11af3d6a=""
          data-v-562f08a9=""
        >
          Object{2}
        </span>
      </div>
      
    </summary>
    <div
      class="content"
      data-v-55c4d015=""
    >
      
      <div
        class="SharedJsonViewerCollapsible__content"
        data-v-11af3d6a=""
      >
        
        <div
          class="SharedJsonViewerObject"
          data-v-4abbcc0e=""
        >
          
          
          <!-- This is a single value, we display it as plain key/value text -->
          
          <!-- This is a an object -->
          <details
            class="SharedCollapsible"
            data-v-11af3d6a=""
            data-v-4abbcc0e=""
            data-v-55c4d015=""
            disabled="false"
          >
            <summary
              class="SharedCollapsible__summary"
              data-v-55c4d015=""
            >
              <!--v-if-->
              
              <div
                class="SharedJsonViewerCollapsible__title"
                data-v-11af3d6a=""
              >
                <span
                  data-v-11af3d6a=""
                >
                  array
                </span>
                <span
                  data-v-11af3d6a=""
                  data-v-562f08a9=""
                >
                  Array[3]
                </span>
              </div>
              
            </summary>
            <div
              class="content"
              data-v-55c4d015=""
            >
              
              <div
                class="SharedJsonViewerCollapsible__content"
                data-v-11af3d6a=""
              >
                
                
              </div>
              
            </div>
          </details>
          
          
          
          <!-- This is a single value, we display it as plain key/value text -->
          
          <!-- This is a an object -->
          <details
            class="SharedCollapsible"
            data-v-11af3d6a=""
            data-v-4abbcc0e=""
            data-v-55c4d015=""
            disabled="false"
          >
            <summary
              class="SharedCollapsible__summary"
              data-v-55c4d015=""
            >
              <!--v-if-->
              
              <div
                class="SharedJsonViewerCollapsible__title"
                data-v-11af3d6a=""
              >
                <span
                  data-v-11af3d6a=""
                >
                  nested
                </span>
                <span
                  data-v-11af3d6a=""
                  data-v-562f08a9=""
                >
                  Object{2}
                </span>
              </div>
              
            </summary>
            <div
              class="content"
              data-v-55c4d015=""
            >
              
              <div
                class="SharedJsonViewerCollapsible__content"
                data-v-11af3d6a=""
              >
                
                
              </div>
              
            </div>
          </details>
          
          
          
        </div>
        
      </div>
      
    </div>
  </details>
  
  <!--v-if-->
  
</div>
`;

exports[`SharedJsonViewer > should render an array 1`] = `
<div
  data-v-app=""
>
  
  
  <details
    class="SharedCollapsible"
    data-v-11af3d6a=""
    data-v-55c4d015=""
    disabled="false"
  >
    <summary
      class="SharedCollapsible__summary"
      data-v-55c4d015=""
    >
      <!--v-if-->
      
      <div
        class="SharedJsonViewerCollapsible__title"
        data-v-11af3d6a=""
      >
        <!--v-if-->
        <span
          data-v-11af3d6a=""
          data-v-562f08a9=""
        >
          Array[3]
        </span>
      </div>
      
    </summary>
    <div
      class="content"
      data-v-55c4d015=""
    >
      
      <div
        class="SharedJsonViewerCollapsible__content"
        data-v-11af3d6a=""
      >
        
        <div
          class="SharedJsonViewerObject"
          data-v-4abbcc0e=""
        >
          
          
          <!-- This is a single value, we display it as plain key/value text -->
          <div
            class="SharedJsonViewerObject__value"
            data-v-4abbcc0e=""
          >
            <span
              data-v-4abbcc0e=""
            >
              0
            </span>
            : 
            <span
              class="SharedJsonViewerValue"
              data-v-4abbcc0e=""
              data-v-e1f0c935=""
            >
              1
            </span>
          </div>
          
          
          <!-- This is a single value, we display it as plain key/value text -->
          <div
            class="SharedJsonViewerObject__value"
            data-v-4abbcc0e=""
          >
            <span
              data-v-4abbcc0e=""
            >
              1
            </span>
            : 
            <span
              class="SharedJsonViewerValue"
              data-v-4abbcc0e=""
              data-v-e1f0c935=""
            >
              2
            </span>
          </div>
          
          
          <!-- This is a single value, we display it as plain key/value text -->
          <div
            class="SharedJsonViewerObject__value"
            data-v-4abbcc0e=""
          >
            <span
              data-v-4abbcc0e=""
            >
              2
            </span>
            : 
            <span
              class="SharedJsonViewerValue"
              data-v-4abbcc0e=""
              data-v-e1f0c935=""
            >
              3
            </span>
          </div>
          
          
        </div>
        
      </div>
      
    </div>
  </details>
  
  <!--v-if-->
  
</div>
`;
